### PR TITLE
build: BAZEL to SBT migration (DEV-508)

### PR DESCRIPTION
**Pull request by mpro7**
resolves DEV-508

### CODE REVIEW

File: project/Dependencies.scala [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+        rdf4jRuntime,
+        scallop
+    )
+
+    val webapiTestAndITLibraryDependencies = l ++= Seq[sbt.ModuleID](
+        TestBinaries.gatlingHighcharts,
+        TestBinaries.gatlingTestFramework,
+        TestBinaries.scalaTest
+    )
+}
```
Comment (BalduinLandolt) [inline project/Dependencies.scala:261]: please add "new line at end of file" to your VS Code settings, if you haven't done so already :)

File: webapi/scripts/expected-client-test-data.txt [hunk 2]
```diff
@@ -14,6 +14,8 @@ test-data/admin/groups/get-groups-response.json
 test-data/admin/groups/update-group-request.json
 test-data/admin/groups/update-group-response.json
 test-data/admin/lists/
+test-data/admin/lists/add-child-to-second-child-request.json
+test-data/admin/lists/add-child-to-second-child-response.json
```
Comment (BalduinLandolt) [inline webapi/scripts/expected-client-test-data.txt:18]: why has this changed in this PR?

File: webapi/src/test/scala/org/knora/webapi/e2e/ClientTestDataCollector.scala [hunk 3]
```diff
@@ -8,53 +8,51 @@ package org.knora.webapi.e2e
 import org.knora.webapi.exceptions.TestConfigurationException
 import org.knora.webapi.settings.KnoraSettingsImpl
 import redis.clients.jedis.{Jedis, JedisPool, JedisPoolConfig}
```
Comment (irinaschubert) [inline webapi/src/test/scala/org/knora/webapi/e2e/ClientTestDataCollector.scala:10]: I see that you deleted redis related things, do we still need the redis library here?

File: webapi/src/main/scala/org/knora/webapi/app/ApplicationActor.scala [hunk 4]
```diff
@@ -521,15 +516,12 @@ class ApplicationActor extends Actor with Stash with LazyLogging with AroundDire
 
   override def postStop(): Unit = {
     super.postStop()
-    logger.info("ApplicationActor - shutdown in progress, initiating post stop cleanup. Bye!")
-
-    if (knoraSettings.prometheusEndpoint) {
-      // Stop Kamon monitoring
-      Kamon.stopModules()
-    }
+    logger.info("ApplicationActor - ... shutdown in progress, initiating post stop cleanup...")
```
Comment (BalduinLandolt) [inline webapi/src/main/scala/org/knora/webapi/app/ApplicationActor.scala:519]: I have no experience with logging in Scala, but from the Python world, it seems to me, that the logger should know where it is, so it should imo. not be necessary to tell it explicitly that we are in `ApplicationActor` here. Maybe we can discuss this some time - I'm generally a bit unclear about our logging conventions, and what may need change there as well.

File: webapi/src/test/scala/org/knora/webapi/e2e/admin/lists/OldListsRouteADMFeatureE2ESpec.scala [hunk 5]
```diff
@@ -1038,7 +1038,7 @@ class OldListsRouteADMFeatureE2ESpec
           TestDataFileContent(
```
Comment (irinaschubert) [inline webapi/src/test/scala/org/knora/webapi/e2e/admin/lists/OldListsRouteADMFeatureE2ESpec.scala:1038]: Do we still need this whole OldList... things?

File: webapi/src/test/scala/org/knora/webapi/e2e/http/ServerVersionE2ESpec.scala [hunk 6]
```diff
@@ -28,7 +28,7 @@ class ServerVersionE2ESpec extends E2ESpec(ServerVersionE2ESpec.config) {
 
   "The Server" should {
 
-    "return the custom 'Server' header with every response" in {
+    "return the custom 'Server' header with every response" ignore {
```
Comment (irinaschubert) [inline webapi/src/test/scala/org/knora/webapi/e2e/http/ServerVersionE2ESpec.scala:31]: Why do we ignore this?

File: webapi/src/test/scala/org/knora/webapi/http/version/BuildInfoSpec.scala [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.knora.webapi.http.version
+
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpecLike
+
+/**
+ * This spec is used to test 'ListAdminMessages'.
```
Comment (irinaschubert) [inline webapi/src/test/scala/org/knora/webapi/http/version/BuildInfoSpec.scala:12]: I think this docstring is wrong...

File: webapi/src/test/scala/org/knora/webapi/e2e/http/ServerVersionE2ESpec.scala [hunk 8]
```diff
@@ -28,7 +28,7 @@ class ServerVersionE2ESpec extends E2ESpec(ServerVersionE2ESpec.config) {
...
       val request = Get(baseApiUrl + s"/admin/projects")
```
Comment (BalduinLandolt) [inline webapi/src/test/scala/org/knora/webapi/e2e/http/ServerVersionE2ESpec.scala:32]: why is this test ignored?

**Review by irinaschubert — APPROVED** [review]
I don't feel like I have much to say but it looks interesting :) Added mostly minor deletion suggestions for uncommented code.

File: .gitignore [hunk 9]
```diff
@@ -1,6 +1,8 @@
 /.project
 .ijwb
 .vscode
+.bloop
+.bsp
```
Comment (BalduinLandolt) [inline .gitignore:5]: duplication below

File: webapi/src/test/scala/org/knora/webapi/e2e/v2/ValuesRouteV2E2ESpec.scala [hunk 10]
```diff
@@ -968,17 +968,6 @@ class ValuesRouteV2E2ESpec extends E2ESpec {
...
-        TestDataFileContent(
-          filePath = TestDataFilePath(
-            directoryPath = clientTestDataPath,
-            filename = "create-int-value-with-custom-UUID-request",
-            fileExtension = "json"
-          ),
-          text = jsonLDEntity
-        )
-      )
-
```
Comment (BalduinLandolt) [inline webapi/src/test/scala/org/knora/webapi/e2e/v2/ValuesRouteV2E2ESpec.scala:981]: why is this deleted?

File: build.sbt [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+import sbt.Keys.version
+
+import scala.language.postfixOps
+import scala.sys.process.Process
+
+//////////////////////////////////////
+// GLOBAL SETTINGS
+//////////////////////////////////////
+
+lazy val aggregatedProjects: Seq[ProjectReference] = Seq(webapi)
```
Comment (BalduinLandolt) [inline build.sbt:14]: do we need this aggregation, if it's only one element in it?

File: build.sbt [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+lazy val root: Project = Project(id = "root", file("."))
+  .aggregate(aggregatedProjects: _*)
+  .enablePlugins(GitVersioning, GitBranchPrompt)
+  .settings(Dependencies.Versions)
+  .settings(
+    // values set for all sub-projects
+    // These are normal sbt settings to configure for release, skip if already defined
+    Global / onChangedBuildSource := ReloadOnSourceChanges,
+    ThisBuild / licenses := Seq("AGPL-3.0" -> url("https://opensource.org/licenses/AGPL-3.0")),
```
Comment (BalduinLandolt) [inline build.sbt:31]: license should be Apache, right?

File: build.sbt [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+    ThisBuild / homepage := Some(url("https://github.com/dhlab-basel/Knora")),
+    ThisBuild / scmInfo := Some(
+      ScmInfo(url("https://github.com/dhlab-basel/Knora"), "scm:git:git@github.com:dhlab-basel/Knora.git")
+    ),
```
Comment (BalduinLandolt) [inline build.sbt:35]: those URLs are outdated, we're dasch not dhlab-basel now

**Review by BalduinLandolt — APPROVED** [review]
certainly the license and URLs need to be adjusted in the build file. Apart from that just minor comments and questions.  
(Also, sorry that I didn't bundle most comments with the review!)

Overall I like SBT much better than Bazel :)

File: webapi/scripts/expected-client-test-data.txt (hunk 2 above)
Comment (mpro7) [reply webapi/scripts/expected-client-test-data.txt:18]: As far as I see it is just filename adjustment to match the test name.

File: project/Dependencies.scala (hunk 1 above)
Comment (mpro7) [reply project/Dependencies.scala:261]: Good point. I think we should also share the core vscode settings.

File: webapi/src/test/scala/org/knora/webapi/e2e/v2/ValuesRouteV2E2ESpec.scala (hunk 10 above)
Comment (mpro7) [reply webapi/src/test/scala/org/knora/webapi/e2e/v2/ValuesRouteV2E2ESpec.scala:981]: Interesting finding, I can't even spot it in the git blame who did it 😄

File: build.sbt (hunk 11 above)
Comment (mpro7) [reply build.sbt:14]: I think we are going to add more projects in new steps, therefore I leave it.

File: webapi/src/test/scala/org/knora/webapi/e2e/v2/ValuesRouteV2E2ESpec.scala (hunk 10 above)
Comment (mpro7) [reply webapi/src/test/scala/org/knora/webapi/e2e/v2/ValuesRouteV2E2ESpec.scala:981]: Ok found it - @subotic why did you delete it? https://github.com/dasch-swiss/dsp-api/pull/2002/commits/5d66fa9939c86a9d5dcb95e6e31b0f1ac95817fa

File: project/plugins.sbt [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+addSbtPlugin("com.lightbend.sbt" % "sbt-javaagent" % "0.1.5")
+addSbtPlugin("io.gatling" % "gatling-sbt" % "2.2.2")
+addSbtPlugin("org.scoverage" % "sbt-scoverage" % "1.6.1")
+addSbtPlugin("com.codacy" % "sbt-codacy-coverage" % "1.3.15")
+addSbtPlugin("com.eed3si9n" % "sbt-buildinfo" % "0.11.0")
+
+// addSbtPlugin("com.timushev.sbt" % "sbt-updates" % "0.6.1")
+// addSbtPlugin("com.github.cb372" % "sbt-explicit-dependencies" % "0.2.16")
+// addDependencyTreePlugin
```
[the remark this answers is not in the document]
Comment (mpro7) [reply project/plugins.sbt:23]: These are going to be used in next tasks.

File: third_party/dependencies.bzl [hunk 15]
```diff
@@ -90,7 +90,7 @@ def dependencies():
             "org.rogach:scallop_2.13:3.5.1",
             "com.google.gwt:gwt-servlet:2.8.0",
             "net.sf.saxon:Saxon-HE:9.9.0-2",
-            "org.scala-lang.modules:scala-xml_2.13:1.1.1",
+            # "org.scala-lang.modules:scala-xml_2.13:1.1.1",
```
[the remark this answers is not in the document]
Comment (mpro7) [reply third_party/dependencies.bzl:93]: In next few steps of migration this whole fole going to be removed.

File: webapi/src/test/scala/org/knora/webapi/e2e/admin/lists/OldListsRouteADMFeatureE2ESpec.scala (hunk 5 above)
Comment (mpro7) [reply webapi/src/test/scala/org/knora/webapi/e2e/admin/lists/OldListsRouteADMFeatureE2ESpec.scala:1038]: Yes, this is the only one listsADM implementation, the "new" one was deleted.

File: webapi/src/test/scala/org/knora/webapi/e2e/http/ServerVersionE2ESpec.scala (hunk 6 above)
Comment (mpro7) [reply webapi/src/test/scala/org/knora/webapi/e2e/http/ServerVersionE2ESpec.scala:31]: Forgot to bring it back.

File: webapi/src/test/scala/org/knora/webapi/e2e/admin/lists/OldListsRouteADMFeatureE2ESpec.scala (hunk 5 above)
Comment (irinaschubert) [reply webapi/src/test/scala/org/knora/webapi/e2e/admin/lists/OldListsRouteADMFeatureE2ESpec.scala:1038]: Ah, OK. Shouldn't we rename it then to just "List..." instead of "OldList..."?

File: .github/workflows/main.yml [hunk 16]
```diff
@@ -65,113 +79,24 @@ jobs:
           architecture: 'x64' # defaults to 'x64'
       - name: install requirements
         run: sudo apt-get install ca-certificates-java expect redis-tools unzip
```
[the remark this answers is not in the document]
Comment (mpro7) [reply .github/workflows/main.yml:81]: Most probably not, but there is a separate task to clean it up.

File: webapi/src/test/scala/org/knora/webapi/CoreSpec.scala [hunk 17]
```diff
@@ -141,8 +141,9 @@ abstract class CoreSpec(_system: ActorSystem)
     // memusage()
   }
 
-  final override def afterAll(): () =
-    appActor ! AppStop()
+  final override def afterAll(): Unit =
+    // appActor ! AppStop()
```
[the remark this answers is not in the document]
Comment (subotic) [reply webapi/src/test/scala/org/knora/webapi/CoreSpec.scala:145]: no, removed

File: .github/workflows/main.yml (hunk 16 above)
[the remark this answers is not in the document]
Comment (subotic) [reply .github/workflows/main.yml:81]: removed

File: webapi/src/test/scala/org/knora/webapi/E2ESpec.scala [hunk 18]
```diff
@@ -110,7 +114,8 @@ class E2ESpec(_system: ActorSystem)
 
   override def afterAll(): Unit =
     /* Stop the server when everything else has finished */
-    appActor ! AppStop()
+    // appActor ! AppStop()
```
[the remark this answers is not in the document]
Comment (subotic) [reply webapi/src/test/scala/org/knora/webapi/E2ESpec.scala:117]: no, removed

File: webapi/src/test/scala/org/knora/webapi/ITKnoraLiveSpec.scala [hunk 19]
```diff
@@ -112,7 +113,8 @@ class ITKnoraLiveSpec(_system: ActorSystem)
 
   override def afterAll(): Unit =
     /* Stop the server when everything else has finished */
-    appActor ! AppStop()
+    // appActor ! AppStop()
```
[the remark this answers is not in the document]
Comment (subotic) [reply webapi/src/test/scala/org/knora/webapi/ITKnoraLiveSpec.scala:116]: no, removed

File: webapi/src/test/scala/org/knora/webapi/e2e/v2/ValuesRouteV2E2ESpec.scala (hunk 10 above)
Comment (subotic) [reply webapi/src/test/scala/org/knora/webapi/e2e/v2/ValuesRouteV2E2ESpec.scala:981]: because it was in the "do not create an integer value if the custom UUID is not part of the custom IRI" test. the "create an integer value with a custom creation date" is the next test (same file name). So nothing changes by removing it.
